piptools/writer.py

```diff
@@ -104,12 +125,24 @@ def write_flags(self):
     def _iter_lines(
         self,
         results,
-        unsafe_requirements,
-        reverse_dependencies,
-        primary_packages,
-        markers,
-        hashes,
+        unsafe_requirements=None,
+        reverse_dependencies=None,
+        primary_packages=None,
+        markers=None,
+        hashes=None,
     ):
+        # default values
+        unsafe_requirements = unsafe_requirements or []
+        reverse_dependencies = reverse_dependencies or {}
+        primary_packages = primary_packages or []
+        markers = markers or {}
+        hashes = hashes or {}
+
+        # Check for unhashed or unpinned packages if at least one package does have
+        # hashes, which will trigger pip install's --require-hashes mode.
+        warn_uninstallable = False
+        has_hashes = hashes and any(hash for hash in hashes.values())
```

Comment:
What does the `any()` do here really, is it necessary?

Reply:
The idea is to only warn if the file is actually going to be rejected by pip, meaning some requirements have hashes and some don't.

The `any()` means we won't warn if a file contains *only* non-hashable requirements, like this:

```
$ cat requirements.in
-e git+https://github.com/pallets/click@master#egg=click
$ pip-compile requirements.in --generate-hashes
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --generate-hashes requirements.in
#
-e git+https://github.com/pallets/click@master#egg=click
```

`hashes` in that situation will be `{<InstallRequirement object: click from git+https://github.com/pallets/click@master#egg=click (from -r requirements.in (line 1)) editable=True>: set()}`, so we have to check the values.

It seems better not to warn in this case because people might be intentionally splitting out editable packages into separate files -- it's been [recommended by pip maintainers](https://github.com/pypa/pip/issues/4995#issuecomment-360828403) to have a separate requirements.txt for editable requirements if you're using hashes. So users might run pip-compile with the same settings on a folder full of requirements files and not need to see warnings in this case.